From b7044c52cf1673fd3fc7ccd0f491993b3fb81aac Mon Sep 17 00:00:00 2001
From: Zhan Lou <zhan.lou@cixtech.com>
Date: Thu, 7 Aug 2025 20:35:49 +0800
Subject: [PATCH] DPTSW-18843: v4l2: deprioritize RGB in decoder src pad caps

Deprioritize RGB format in decoder src pad caps as we prefer YUV formats
for better performance and lower memory footprint.

Signed-off-by: Zhan Lou <zhan.lou@cixtech.com>
Change-Id: I2893b3226afe6d9b0d881ece72d7af2251fd30d7

diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index 3035939aa9..1325ef17ae 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -491,6 +491,28 @@ gst_v4l2_video_try_downscaling_caps (GstVideoDecoder * decoder,
   return out_caps;
 }
 
+static void
+gst_v4l2_video_dec_deprioritize_rgb (GstCaps * caps)
+{
+  gint caps_size = gst_caps_get_size (caps);
+  gint rgb_caps_count = 0;
+  gint i;
+
+  if (caps_size < 2)
+    return;
+
+  for (i = 0; i < caps_size; i++) {
+    GstStructure *s = gst_caps_get_structure (caps, i);
+    const gchar *format = gst_structure_get_string (s, "format");
+
+    if (g_str_equal (format, "RGB")) {
+      s = gst_caps_steal_structure (caps, i);
+      gst_caps_append_structure (caps, s);
+      rgb_caps_count++;
+    }
+  }
+}
+
 static gboolean
 gst_v4l2_video_dec_negotiate (GstVideoDecoder * decoder)
 {
@@ -603,7 +625,11 @@ re_negotiate:
   }
   gst_caps_unref (filter);
 
-  GST_DEBUG_OBJECT (self, "Possible decoded caps: %" GST_PTR_FORMAT, caps);
+  /* Deprioritize RGB caps as we prefer YUV for better performance */
+  if (!gst_caps_is_empty (caps))
+    gst_v4l2_video_dec_deprioritize_rgb (caps);
+
+GST_DEBUG_OBJECT (self, "Possible decoded caps: %" GST_PTR_FORMAT, caps);
   if (gst_caps_is_empty (caps)) {
     gst_caps_unref (caps);
     goto not_negotiated;
